csharp/src/SeedLang.Common/BlockPosition.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace SeedLang.Common {
+  // An immutable class to represent a position in a SeedBlock source code.
+  //
+  // Typically a position in a block-style language is the ID of a specified block.
+  public class BlockPosition : IEquatable<BlockPosition> {
+    public string BlockId { get; }
+
+    public BlockPosition(string blockId) {
+      BlockId = blockId;
+    }
+
+    public override int GetHashCode() {
+      return BlockId.GetHashCode();
+    }
+
+    public override string ToString() {
+      return $"Block: {BlockId}";
+    }
+
+    public bool Equals(BlockPosition pos) {
+      return !(pos is null) && BlockId == pos.BlockId;
+    }
+
+    public override bool Equals(object obj) {
+      return (obj is BlockPosition objBlockPosition) && Equals(objBlockPosition);
```

Comment:
The code can be "this.Equals(obj as BlockPosition);" I tried another implement "return false;", which passes all the test cases as well. I think the reason is if obj is the type of BlockPosition, it will call the type specific version. This version will handle all cases of non BlockPosition type. It's true if the BlockPosition is sealed without subclass.

Reply:
Let's follow the "How to define value equality for a type" doc here.

Reply:
In the updated code: class equality (like in class Range) uses "Equals(obj as Range)". struct equality (like in TextPosition) still uses "(obj is TextPosition ...)" since struct types are not nullable.